### ISSUE TITLE
Close the WGPU operation gap in tiers (112 operations)

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#91]
## Problem

WGPU advertises 46 of the 158 backend-executable catalog operations, the widest gap of any backend that actually executes. Unlike CUDA and Metal, WGPU is the portable path: it is what a contributor without an NVIDIA card, and a user on an integrated GPU or in a browser, would use. It is also the only accelerator with automated execution coverage today, through the software adapter.

The 112-operation gap by category:

| Category | Count | Operations |
|---|---:|---|
| Layout and views | 17 | `broadcast_left`, `chunk`, `concat`, `flatten`, `narrow`, `pad`, `pixel_shuffle`, `repeat`, `slice`, `split`, `squeeze`, `stack`, `transpose`, `tril`, `triu`, `unfold`, `unsqueeze` |
| Unary math | 21 | `acos`, `acosh`, `asin`, `asinh`, `atan`, `atanh`, `ceil`, `cos`, `cosh`, `erf`, `floor`, `frac`, `log2`, `log10`, `round`, `rsqrt`, `sign`, `sin`, `sinh`, `tan`, `trunc` |
| Binary and scalar | 13 | `abs_diff`, `add_scalar`, `sub_scalar`, `mul_scalar`, `div_scalar`, `atan2`, `clamp`, `fmod`, `lerp`, `maximum`, `minimum`, `powf`, `remainder` |
| Reduction and statistics | 12 | `argmax`, `argmin`, `argsort`, `cumsum`, `norm`, `std_all`, `std_dim`, `std_keepdim`, `topk`, `var_all`, `var_dim`, `var_keepdim` |
| Comparison and logic | 9 | `cmp_eq`, `cmp_ge`, `cmp_gt`, `cmp_le`, `cmp_lt`, `cmp_ne`, `logical_and`, `logical_not`, `logical_or` |
| Indexing | 7 | `diag`, `embedding`, `gather`, `index_select`, `masked_fill`, `scatter`, `where_cond` |
| Matmul family | 6 | `addmm`, `bmm`, `dot`, `linear`, `outer`, `scaled_dot_product_attention` |
| Loss and regularization | 5 | `bce_with_logits_loss`, `cross_entropy_loss`, `l1_loss`, `mse_loss`, `dropout` |
| Normalization | 5 | `batch_norm`, `group_norm`, `instance_norm`, `layer_norm`, `rms_norm` |
| Creation and readback | 9 | tracked in #89 |
| Conv and pool | 3 | `adaptive_avg_pool2d`, `conv1d`, `conv_transpose2d` |
| Quantization | 4 | `dequantize`, `quantize`, `quantized_matmul`, `to_dtype` |
| Other | 1 | `softmax` |

WGPU also has no normalization at all, and no `softmax`, so it is further from a working transformer than CUDA is.

## Proposed implementation

Tier the work rather than treating 112 operations as one change:

- **Tier 1, portable transformer path:** `softmax`, `layer_norm`, `rms_norm`, `transpose`, `concat`, `slice`, `narrow`, `squeeze`, `unsqueeze`, `flatten`, `masked_fill`, `where_cond`, `tril`, `triu`, `embedding`, `bmm`, `linear`. This is the set that turns WGPU from an arithmetic demo into something that can run an attention block.
- **Tier 2, training:** the five losses, `dropout`, `batch_norm`, `group_norm`, and the backward paths for tier 1.
- **Tier 3, breadth:** the remaining unary, binary, comparison, and statistics operations, generated from one WGSL pointwise template rather than hand-written per operation.

Constraints specific to this backend that should be settled once and recorded, not rediscovered per operation:

- WGSL has no `f64`. Every row must state its dtype set honestly rather than inheriting the CPU set.
- `f16` requires the `shader-f16` capability, which is not universally available; advertise it conditionally on adapter features rather than unconditionally.
- Workgroup size and storage-buffer binding limits differ by adapter, so the reduction and scan kernels need a limit-aware launch configuration rather than a constant.

## Acceptance criteria

- Tier 1 executes on the software adapter in CI and on a native adapter once #82 provides one.
- An attention block runs end to end on WGPU.
- Advertised dtype sets reflect actual WGSL and adapter capability, with no row advertising `f64`.
- `f16` rows are advertised only when the adapter reports `shader-f16`.
- The conformance harness in #83 passes for every advertised row.

## Runtime and driver dependencies

Filed partly against `backend_request.yml`, since the constraints here are backend-architectural rather than per-operation.

WGPU is the portable path: no vendor SDK, no driver-specific runtime, and the only accelerator with automated execution coverage today (lavapipe software adapter on `ubuntu-latest`). It is what a contributor without an NVIDIA card uses, and what a user on integrated graphics or in a browser gets.

Constraints that must be settled **once and recorded**, not rediscovered per operation:

| Constraint | Consequence |
|---|---|
| **WGSL has no `f64`** | Every WGPU row must state its dtype set honestly. **No row may advertise `f64`**, and inheriting the CPU dtype set blindly would do exactly that. |
| **`f16` needs the `shader-f16` capability** | Not universally available. Advertise `f16` rows **conditionally on adapter features**, queried at device creation, not unconditionally at compile time. |
| **Workgroup size and storage-buffer binding limits vary by adapter** | Reduction and scan kernels need a limit-aware launch configuration, not a constant. `maxComputeWorkgroupSizeX` and `maxStorageBuffersPerShaderStage` are the two that bite. |
| **No native atomics on floats** | `scatter_add`-style accumulation (#103) needs a different strategy than CUDA's. Worth knowing before that work starts. |
| **Buffer size limits** | `maxStorageBufferBindingSize` is commonly 128 MiB. Large tensors need chunked dispatch or an explicit refusal. |

## Memory and buffer management model

Existing: `crates/incin-backends/src/wgpu/backend/` owns buffer allocation and `pollster` drives the async submission to a blocking API. This issue does not change that model; it adds operations on top of it. Two notes that affect the tiering:

- **Staging buffers dominate small operations.** Every host-visible readback needs a mapped staging buffer and a queue submission. That is why the host-readback rows in #89 are worth doing once, centrally, rather than per operation.
- **Bind-group layout churn.** Each distinct shader signature is a pipeline. Emitting 100+ operations as 100+ hand-written WGSL files with bespoke bind groups is where this becomes unmaintainable; the tier-3 pointwise template exists to prevent that.

## Proposed API and capability-row change

No new public API. The `Tensor` methods exist; the WGPU capability table gains rows.

```rust
use incin::prelude::*;
type Backend = DefaultBackend;   // with `wgpu`, the WGPU backend

// Tier 1 target: an attention block that runs anywhere, including a software adapter.
let scores = q.bmm(&k_t)?.mul_scalar(scale)?;   // bmm: refusal today
let attn   = scores.softmax(1)?;                // softmax: refusal today
let out    = attn.bmm(&v)?;
let normed = out.layer_norm(&gamma, &beta, 1e-5)?;         // refusal today
```

```rust
// crates/incin-backends/src/capability/tables.rs - proposed shape of the change
pub static WGPU_CAPABILITIES: &[CapabilityRule] = wgpu_descriptor_operations!(
    descriptor_capability_rules,
    elementwise = WGSL_FLOAT_DTYPES,        // proposed: f32, plus f16 iff shader-f16
    normalization_dtypes = F32_ONLY,        // proposed: was absent entirely
    matmul = WGSL_FLOAT_DTYPES,
    // NOTE: no constant here may include f64.
);
```

## Shape and type-system invariants

Shape rules are the catalog's and do not change per backend. What is backend-specific:

- Advertised **dtype sets** must reflect actual WGSL and adapter capability. A row advertising a dtype the shader cannot express is the exact failure `#83` exists to catch.
- Advertised **layouts**: WGPU rows should advertise `contiguous` unless a shader genuinely handles strided input. The CPU rows advertise `CPU_LAYOUTS` (contiguous and strided); copying that constant across would be a false claim.
- Advertised **rank ranges** must match what the shader's bind-group layout actually indexes.

## How this interoperates

- **#89** supplies the creation and host-readback rows, shared across all three accelerators.
- **#83** is what verifies the conditional `f16` advertisement: a row advertised only when the adapter reports `shader-f16` must be skipped, not failed, on an adapter without it, and the harness has to tell those apart.
- **#82** supplies a native-adapter runner, so WGPU execution stops resting solely on lavapipe. Note that the software-adapter job **does** run today, which is why WGPU is the least blocked of the three accelerators.
- **#100** determines whether `where_cond` and `masked_fill`, both in the tier-1 set, are reachable from statically shaped tensors.
- **#101** is the consumer: the tier-1 set is chosen precisely to make an attention block run.
- **`docs/FEATURE_MATRIX.md`** and the `wgpu` feature-contract row record the tier this reaches.

## Alternatives

Treating WGPU as a permanent demo tier and concentrating on CUDA would be cheaper, but gives up the only backend that runs everywhere and the only accelerator currently covered by automated execution. If that is the decision, it should be recorded and the capability document should say so, rather than being the outcome of never scheduling the work.

## Additional context

Relevant code: `crates/incin-backends/src/wgpu/shaders/`, `crates/incin-backends/src/wgpu/backend/`, and the CPU references under `crates/incin-backends/src/cpu/ops/`.

Creation and host-readback rows are tracked separately in #89. Verified by #83.

**Comment by xupremix** [comment]
Recount against `master` today, since #89 landed.

WGPU now advertises **64** of **163** backend-executable catalog operations, not 46 of 158. The gap is **99**, not 112.

The "Creation and readback | 9 | tracked in #89" row of the table above is done: WGPU advertises all eleven creation and readback rows now. The other eighteen came from the catalog growing and from rows widening.

The per-category table is otherwise still the right shape — layout and views, unary math, comparison and logic, and the matmul family are all still missing wholesale — so the tiering plan does not change. The header count does.

Counts derived from `docs/api-site-data.json`, generated from `docs/capabilities.md`. Reproduce with `python3 tools/build-api-site-data.py`.
